src/juice.c

```diff
@@ -147,6 +148,14 @@ JUICE_EXPORT int juice_get_selected_addresses(juice_agent_t *agent, char *local,
 	return JUICE_ERR_SUCCESS;
 }
 
+int juice_set_local_ice_attributes(juice_agent_t *agent, const char *ufrag, const char *pwd)
+{
+	strncpy(agent->local.ice_ufrag, ufrag, sizeof(agent->local.ice_ufrag));
+	strncpy(agent->local.ice_pwd, pwd, sizeof(agent->local.ice_pwd));
```

Comment:
If the local description is changed after gathering started, it could cause a race condition with the agent thread. This function should delegate to a function `agent_set_local_ice_attributes` in `agent.c` which would begin with a check like:
```
	if (agent->conn_impl) {
		JLOG_WARN("Candidates gathering already started");
		return -1;
	}
```

Reply:
Thanks for pointing this out. Is it ok to use strncpy_s? I think it's a bit heavy to use snprintf here.

Reply:
The problem is that strncpy_s is not always available, which is a pain for portability:
> As with all bounds-checked functions, strncpy_s only guaranteed to be available if `__STDC_LIB_EXT1__` is defined by the implementation and if the user defines `__STDC_WANT_LIB_EXT1__` to the integer constant 1 before including <string.h>.

It's just easier and more portable to use something like `snprintf(agent->local.ice_ufrag, sizeof(agent->local.ice_ufrag), "%s", ufrag)`

Reply:
get it